vault/logical_system.go

```diff
@@ -3044,6 +3063,72 @@ func (b *SystemBackend) pathInternalUIResultantACL(ctx context.Context, req *log
 	return resp, nil
 }
 
+func (b *SystemBackend) pathInternalUIFilteredPath(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	path := d.Get("path").(string)
+	if path == "" {
+		return logical.ErrorResponse("path not set"), logical.ErrInvalidRequest
+	}
+	path = sanitizeMountPath(path)
+
+	errResp := logical.ErrorResponse(fmt.Sprintf("preflight capability check returned 403, please ensure client's policies grant access to path %q", path))
+
+	ns, err := namespace.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Load the ACL policies so we can walk the prefix for this mount
+	acl, te, entity, _, err := b.Core.fetchACLTokenEntryAndEntity(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	if entity != nil && entity.Disabled {
+		b.logger.Warn("permission denied as the entity on the token is disabled")
+		return errResp, logical.ErrPermissionDenied
+	}
+	if te != nil && te.EntityID != "" && entity == nil {
+		b.logger.Warn("permission denied as the entity on the token is invalid")
+		return nil, logical.ErrPermissionDenied
+	}
+
+	if !hasMountAccess(ctx, acl, ns.Path+path) {
+		return errResp, logical.ErrPermissionDenied
+	}
+
+	storage, mountEntry := b.Core.router.MatchingStorageAndMountByApiPath(ctx, path)
+	if storage == nil {
+		// Return a permission denied error here so this path cannot be used to
+		// brute force a list of mounts.
+		return errResp, logical.ErrPermissionDenied
+	}
+
+	relpath := strings.TrimPrefix(path, mountEntry.Path)
+	unfiltered, err := storage.List(ctx, relpath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return early if we have list access to the path.
+	newreq := *req
+	newreq.Path = path
+	if acl.AllowOperation(ctx, &newreq, false).Allowed {
+		return logical.ListResponse(unfiltered), nil
+	}
+
+	filtered := make([]string, 0, len(unfiltered))
+	for _, key := range unfiltered {
+		isSubTree := strings.HasSuffix(key, "/")
+		fullPath := ns.Path + filepath.Join(path, key)
```

Comment:
Will you add a test that uses namespaces?

Reply:
Certainly.